File: 3176. Find the Maximum Length of a Good Subsequence I.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        dp = [[1] * (k+1) for i in range(len(nums))]
        answer = 0
        for x in range(k+1):
            for i in range(len(nums)):
                for j in range(i):
                    if nums[i] == nums[j]:
                        dp[i][x] = max(dp[i][x], 1+dp[j][x])
                    elif x > 0:
                        dp[i][x] = max(dp[i][x], 1+dp[j][x-1])
                answer = max(answer, dp[i][x])
        return answer

File: 3176. Find the Maximum Length of a Good Subsequence I.java
class Solution {
    public int maximumLength(int[] nums, int k) {
        int[][] dp = new int[nums.length][k+1];
        for (int[] x : dp) {
            Arrays.fill(x, 1);
        }
        int answer = 0;
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] == nums[j]) {
                    dp[i][0] = Math.max(dp[i][0], 1+dp[j][0]);
                }
            }
            answer = Math.max(answer, dp[i][0]);
        }
        for (int x = 1; x <= k; x++) {
            for (int i = 0; i < nums.length; i++) {
                for (int j = 0; j < i; j++) {
                    if (nums[i] == nums[j]) {
                        dp[i][x] = Math.max(dp[i][x], 1+dp[j][x]);
                    } else {
                        dp[i][x] = Math.max(dp[i][x], 1+dp[j][x-1]);
                    }
                }
                answer = Math.max(answer, dp[i][x]);
            }
        }
        return answer;
    }
}
